numpy/lib/arraypad.py

```diff
@@ -1272,43 +1236,57 @@ def pad(array, pad_width, mode, **kwargs):
     if mode == 'constant':
         for axis, ((pad_before, pad_after), (before_val, after_val)) \
                 in enumerate(zip(pad_width, kwargs['constant_values'])):
-            newmat = _prepend_const(newmat, pad_before, before_val, axis)
-            newmat = _append_const(newmat, pad_after, after_val, axis)
+            if pad_before:
+                newmat = _do_prepend(newmat, _leading_const(newmat, pad_before, before_val, axis), axis)
+            if pad_after:
+                newmat = _do_append(newmat, _trailing_const(newmat, pad_after, after_val, axis), axis)
 
     elif mode == 'edge':
         for axis, (pad_before, pad_after) in enumerate(pad_width):
-            newmat = _prepend_edge(newmat, pad_before, axis)
-            newmat = _append_edge(newmat, pad_after, axis)
+            if pad_before:
+                newmat = _do_prepend(newmat, _leading_edge(newmat, pad_before, axis), axis)
+            if pad_after:
+                newmat = _do_append(newmat, _trailing_edge(newmat, pad_after, axis), axis)
 
     elif mode == 'linear_ramp':
         for axis, ((pad_before, pad_after), (before_val, after_val)) \
                 in enumerate(zip(pad_width, kwargs['end_values'])):
-            newmat = _prepend_ramp(newmat, pad_before, before_val, axis)
-            newmat = _append_ramp(newmat, pad_after, after_val, axis)
+            if pad_before:
+                newmat = _do_prepend(newmat, _leading_ramp(newmat, pad_before, before_val, axis), axis)
+            if pad_after:
+                newmat = _do_append(newmat, _trailing_ramp(newmat, pad_after, after_val, axis), axis)
 
     elif mode == 'maximum':
         for axis, ((pad_before, pad_after), (chunk_before, chunk_after)) \
                 in enumerate(zip(pad_width, kwargs['stat_length'])):
-            newmat = _prepend_max(newmat, pad_before, chunk_before, axis)
-            newmat = _append_max(newmat, pad_after, chunk_after, axis)
+            if pad_before:
+                newmat = _do_prepend(newmat, _leading_max(newmat, pad_before, chunk_before, axis), axis)
+            if pad_after:
+                newmat = _do_append(newmat, _trailing_max(newmat, pad_after, chunk_after, axis), axis)
 
     elif mode == 'mean':
         for axis, ((pad_before, pad_after), (chunk_before, chunk_after)) \
                 in enumerate(zip(pad_width, kwargs['stat_length'])):
-            newmat = _prepend_mean(newmat, pad_before, chunk_before, axis)
-            newmat = _append_mean(newmat, pad_after, chunk_after, axis)
+            if pad_before:
+                newmat = _do_prepend(newmat, _leading_mean(newmat, pad_before, chunk_before, axis), axis)
+            if pad_after:
+                newmat = _do_append(newmat, _trailing_mean(newmat, pad_after, chunk_after, axis), axis)
```

Comment:
As a next step, an easy optimization here would be:
```python
parts = (newmat,)
if pad_before:
    parts = (_leading_mean(...),) + parts
if pad_after:
    parts = parts + (_trailing_mean(...),)
if len(parts) > 1:
    newmat = np.concatenate(parts)
```

notably this would also fix #11216